components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java

```diff
@@ -452,93 +452,128 @@ private String getOIDCAuthzEndpoint(Map<String, String> authenticatorProperties)
 
-            if (StringUtils.isNotBlank(idToken)) {
-                jsonObject = getIdTokenClaims(context, idToken);
-                if (jsonObject == null) {
-                    String errorMessage = ErrorMessages.DECODED_JSON_OBJECT_IS_NULL.getMessage();
-                    if (log.isDebugEnabled()) {
-                        log.debug(errorMessage);
-                    }
-                    throw new AuthenticationFailedException(ErrorMessages.DECODED_JSON_OBJECT_IS_NULL.getCode(),
-                            errorMessage);
-                }
+            jsonObject.entrySet().stream()
+                    .filter(entry -> !ArrayUtils.contains(NON_USER_ATTRIBUTES, entry.getKey()))
+                    .forEach(entry -> buildClaimMappings(claims, entry, attributeSeparator));
 
-                String idpName = context.getExternalIdP().getIdPName();
-                String sidClaim = (String) jsonObject.get(OIDCAuthenticatorConstants.Claim.SID);
-                if (StringUtils.isNotBlank(sidClaim) && StringUtils.isNotBlank(idpName)) {
-                    // Add 'sid' claim into authentication context, to be stored in the UserSessionStore for
-                    // single logout.
-                    context.setProperty(FEDERATED_IDP_SESSION_ID + idpName, sidClaim);
-                }
-
-                if (log.isDebugEnabled() && IdentityUtil
-                        .isTokenLoggable(IdentityConstants.IdentityTokens.USER_ID_TOKEN)) {
-                    log.debug("Retrieved the User Information:" + jsonObject);
-                }
+            authenticatedUser = AuthenticatedUser
+                    .createFederateAuthenticatedUserFromSubjectIdentifier(authenticatedUserId);
+        } else {
+            if (log.isDebugEnabled()) {
+                log.debug("The IdToken is null");
+            }
+            authenticatedUser = AuthenticatedUser.createFederateAuthenticatedUserFromSubjectIdentifier(
+                    getAuthenticateUser(context, jsonObject, oAuthResponse));
+        }
+        claims.putAll(getSubjectAttributes(oAuthResponse, authenticatorProperties));
+        authenticatedUser.setUserAttributes(claims);
+        context.setSubject(authenticatedUser);
+    }
 
-                String authenticatedUserId = getAuthenticatedUserId(context, oAuthResponse, jsonObject);
-                String attributeSeparator = getMultiAttributeSeparator(context, authenticatedUserId);
+    /**
+     * Retrieves or maps the ID token according to the flow supported by the authenticator
+     * @param context           AuthenticationContext
+     * @param request           HttpServletRequest
+     * @param oAuthResponse     OAuthClientResponse
+     * @return The ID token
+     */
+    protected String mapIdToken(AuthenticationContext context, HttpServletRequest request, OAuthClientResponse oAuthResponse) {
+        return oAuthResponse.getParam(OIDCAuthenticatorConstants.ID_TOKEN);
+    }
```

Comment:
Shall we fix formatting errors in this block?

Reply:
Fixed 5de101c26645ca300637dbbb70739df79837a218